### PR TITLE
Implement Certificate Revocation List (3)

### PR DESCRIPTION
**Pull request by Danielius1922**

### CODE REVIEW

File: pkg/security/certManager/general/certManager.go [hunk 1]
```diff
@@ -382,3 +385,51 @@ func (a *CertManager) onFileChange(event fsnotify.Event) {
...
+			continue
+		}
+		if pkgX509.IsRevoked(certificate, crl) {
+			return fmt.Errorf("certificate(%s) was revoked by CRL(%s)", "", crl.Issuer.String())
+		}
+		return nil
+
+	}
+	return errors.New("failed to verify certificate by CRL")
+}
```
[the remark this answers is not in the document]
Comment (Danielius1922) [reply pkg/security/certManager/general/certManager.go]: This is the currently accepted solution, once a CRL is downloaded, we use it for verification and ignore the other distribution points.

File: pkg/net/coap/service/udpServer.go [hunk 2]
```diff
@@ -30,37 +34,44 @@ func (s *dtlsServer) Serve() error {
...
 	s.coapServer.Stop()
-	s.closeListener()
+	s.listener.close()
 	return nil
 }
 
+type udpListerner struct {
+	*net.UDPConn
+	close func()
+}
```
[the remark this answers is not in the document]
Comment (Danielius1922) [reply pkg/net/coap/service/udpServer.go]: let's keep it consistent with tcpListener, which uses a close function member

File: pkg/net/coap/service/udpServer.go [hunk 3]
```diff
@@ -18,10 +18,14 @@ import (
 	"go.opentelemetry.io/otel/trace"
 )
 
+type dtlsListerner struct {
+	coapDtlsServer.Listener
+	close func()
+}
```
[the remark this answers is not in the document]
Comment (Danielius1922) [reply pkg/net/coap/service/udpServer.go]: let's keep it consistent with tcpListener, which uses a close function member

**Comment by Danielius1922** [comment]
@coderabbitai review

File: pkg/security/certManager/general/certManager.go [hunk 4]
```diff
@@ -390,3 +393,47 @@ func (a *CertManager) onFileChange(event fsnotify.Event) {
...
+	}
+	return x509.ParseRevocationList(respBody)
+}
+
+func (a *CertManager) VerifyByCRL(ctx context.Context, certificate *x509.Certificate, cdps []string) error {
+	if !a.config.CRL.Enabled {
+		return nil
+	}
+	for _, dp := range cdps {
+		crl, err := a.downloadCRL(ctx, dp)
```
Comment (jkralik) [inline pkg/security/certManager/general/certManager.go]: what about caching ? this will be too expenses to do it for 10k devices

**Review by jkralik — COMMENTED** [review]
For certificate authority you need to also override CRL verification because for endpoint that points to itself. In this case you look directly to database.

**Comment by Danielius1922** [comment]
@coderabbitai review

File: pkg/security/certManager/general/crl_cache.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	}()
+	respBody, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code %v while downloading CRL from %s", resp.StatusCode, distributionPoint)
+	}
+	return x509.ParseRevocationList(respBody)
+}
```
[the remark this answers is not in the document]
Comment (Danielius1922) [reply pkg/security/certManager/general/crl_cache.go]: How would you validate it?

File: pkg/security/x509/verify.go [hunk 6]
```diff
@@ -58,3 +81,63 @@ func Verify(certificates []*x509.Certificate, certificateAuthorities []*x509.Cer
...
+	certificate := chain[0]
+	intermediateCAPool := x509.NewCertPool()
+	for i := 1; i < len(chain); i++ {
+		intermediateCAPool.AddCert(chain[i])
+	}
+	_, err := certificate.Verify(x509.VerifyOptions{
+		Roots:         capool,
+		Intermediates: intermediateCAPool,
+		CurrentTime:   time.Now(),
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
```
Comment (jkralik) [inline pkg/security/x509/verify.go:104]: If server verifies then client must be there. If client verifies server must be there.

**Review by jkralik — APPROVED** [review]

**Pull request merged**